test_project/test_app/tests/test_storage.py

```diff
@@ -198,6 +215,111 @@ def test_remove_intermediate_results_suggest(self):
+
+    def test_store_facet_data(self):
+        """
+        Storing saves facet data
+        """
+        aapl = Stock.objects.get(symbol='AAPL')
+        provider = FacetedStockAutocompleteProvider(aapl)
+        provider.store()
+
+        provider_name = provider.get_provider_name()
+        # the FacetedStockAutocompleteProvider get_facets is set to ['sector']
+        facet_set_name = base.FACET_SET_BASE_NAME % (provider_name, 'sector', 'Technology',)
+        set_length = self.redis.zcard(facet_set_name)
+        self.assertEqual(set_length, 1)
+
+        facet_map_name = base.FACET_MAP_BASE_NAME % (provider_name,)
+        facet_data = provider._deserialize_data(self.redis.hget(facet_map_name, aapl.id))
+        self.assertEqual(facet_data, [{'key': 'sector', 'value': aapl.sector}])
+
+    def test_second_store_removes_old_facet_data(self):
+        """
+        Store removes outdated facet data and updates mapping
+        """
+        aapl = Stock.objects.get(symbol='AAPL')
+        provider = FacetedStockAutocompleteProvider(aapl)
+        provider.store()
+
+        provider_name = provider.get_provider_name()
+        # the FacetedStockAutocompleteProvider get_facets is set to ['sector']
+        facet_set_name = base.FACET_SET_BASE_NAME % (provider_name, 'sector', 'Technology',)
+        set_length = self.redis.zcard(facet_set_name)
+        self.assertEqual(set_length, 1)
+
+        facet_map_name = base.FACET_MAP_BASE_NAME % (provider_name,)
+        facet_data = provider._deserialize_data(self.redis.hget(facet_map_name, aapl.id))
+        self.assertEqual(facet_data, [{'key': 'sector', 'value': aapl.sector}])
+
+        aapl.sector = 'Healthcare'
+        aapl.save()
+        provider.store()
+        # make sure the old key was removed after the second store call
+        set_length = self.redis.zcard(facet_set_name)
+        self.assertEqual(set_length, 0)
+        facet_set_name = base.FACET_SET_BASE_NAME % (provider_name, 'sector', 'Healthcare',)
+        set_length = self.redis.zcard(facet_set_name)
+        self.assertEqual(set_length, 1)
+
+        facet_data = provider._deserialize_data(self.redis.hget(facet_map_name, aapl.id))
+        self.assertEqual(facet_data, [{'key': 'sector', 'value': 'Healthcare'}])
+
+    def test_remove_facet_data(self):
+        """
+        Remove takes care of deleting facet data
+        """
+        aapl = Stock.objects.get(symbol='AAPL')
+        provider = FacetedStockAutocompleteProvider(aapl)
+        provider.store()
+        provider.remove()
+
+        provider_name = provider.get_provider_name()
+        # the FacetedStockAutocompleteProvider get_facets is set to ['sector']
+        facet_set_name = base.FACET_SET_BASE_NAME % (provider_name, 'sector', 'Technology',)
+        set_length = self.redis.zcard(facet_set_name)
+        self.assertEqual(set_length, 0)
+
+        facet_map_name = base.FACET_MAP_BASE_NAME % (provider_name,)
+        keys = self.redis.keys(facet_map_name)
```

Comment:
hmembers instead of keys?.... if you haven't noticed by now i only like using keys when no other command does what we need, for example when doing patterned searches like "djac.stock.*" and so forth